### ISSUE TITLE
How can we apply data parallelism (single machine multiple GPU) training to this NMT model?

### ISSUE THREAD

**Issue by bshao001** [tensorflow/nmt#141]
I know that model parallelism is implemented in this tutorial. However, I assume the advantage of doing that is mainly for large models. To me, I don't see any training speed improvement for a 2-layer model. I tried 4-layer model, it may have some speed improvement. In order to train a 2-layer model with a large amount of training data, I really want to use data parallelism. However, due to the special way to data feed, I would like to ask:

Traditionally, we multiply the batch size, and then split the training batch for each of the duplicated model running on each GPU. For this model, when any component in self.iterator object is evaluated (run in a tf session), the next batch is automatically generated. Also, the iterator is already in the model graph. I assume, we don't need to multiply the batch size, and then split the training batch any more. Each duplicated model is connected to the data source already. Then, how can we manage to let each duplicated instance (model running on different GPU) to get a difference part of the training sample in one epoch? 

For example, we have two GPUs, and two thousand pairs of training samples. How can we let the first duplicated model running on GPU:0 to get one thousand pairs, while the GPU:1 get the other one thousand pairs?

Any suggestions, advises or ideas would be greatly appreciated! If I understand this incorrectly, please correct as well. Thanks again!

**Comment by bshao001** [comment]
Another related question is: Should we first average the gradients from different duplicated models, then apply clipping to the means, or should we clipping the gradients first, then average them. It looks to me both work. But if you have any existing experience or knowledge over this, please do guide me on that. Thanks again.

**Comment by bshao001** [comment]
To make the first question more specific, should we make a list of iterators (the number of iterators equals to the number of GPUs) or just use a single iterator as in this tutorial? I am wondering if we pass the same iterator to multiple copies of the same model running on multiple GPUs, when get_next() is invoked by different copies, it may work as well. Please advise. Thanks.
